src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/Pipe.cs

```diff
@@ -447,7 +447,30 @@ internal void CompleteWriter(Exception? exception)
 
         internal void AdvanceReader(in SequencePosition consumed)
         {
-            AdvanceReader(consumed, consumed);
+            // If the reader is completed
+            if (_readerCompletion.IsCompleted)
+            {
+                ThrowHelper.ThrowInvalidOperationException_NoReadingAllowed();
+            }
+
+            long examinedIndex = consumed.GetInteger();
+            BufferSegment? examinedSegment = (BufferSegment?)consumed.GetObject();
+            if (examinedSegment != null &&
+                // Avoid the lock if we're examining the entire segment, don't need to look at the last examined index in that case
+                examinedSegment.Length - examinedIndex > 0)
```

Comment:
How are we sure we don't need to look at `_lastExaminedIndex` in the case `((BufferSegment)consumed.GetObject()).Length == consumed.GetInteger()`? The previous `AdvanceTo` call's `examined` argument could have pointed to a subsequent `BufferSegment`.

Isn't the point of this change to guarantee that as long as the call wouldn't unconsume anything we will never throw for unexamining anything either considering the `examined` parameter was left unspecified?

Reply:
> Isn't the point of this change to guarantee that as long as the call wouldn't unconsume anything we will never throw for unexamining anything either considering the examined parameter was left unspecified?

The point is to use the greater of the two examined indexes. Either the current internal one or the `consumed` index which we're using as examined. There should be no un-examining or un-consuming with this change.

Reply:
Maybe a phrased this poorly. I understand that this PR doesn't unexamine or unconsume anything, but it does prevent `AdvanceTo(consumed)` from throwing in a scenario where we previously interpreted it to be a request to unexamine data which we have never allowed. I understand that this PR avoids any actual unexamination by using the previous `examined` cursor, but [I think it could lead to stalling](https://github.com/dotnet/runtime/pull/107360#discussion_r1752329868) in cases where it would have previously thrown.

Removing the `examinedBytes < 0` check and not throwing seems simpler and more robust against accidental stalling. I think that still serves the ultimate point of this change which is to not throw in this scenario.

Reply:
I don't understand how this change leads to stalling unless you're referring to how this opens the door for future changes to accidentally introduce stalling? Is removing the `examinedBytes < 0` check in addition to the current changes? Or are you saying we should allow un-examining?